volume/modules/local/grpquota.go

```diff
@@ -0,0 +1,313 @@
+package local
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/alibaba/pouch/pkg/bytefmt"
+	"github.com/alibaba/pouch/pkg/exec"
+
+	"github.com/sirupsen/logrus"
+)
+
+// GrpQuota represents grp quota.
+type GrpQuota struct {
+}
+
+// StartQuotaDriver is used to start quota driver.
+func (quota *GrpQuota) StartQuotaDriver(dir string) (string, error) {
+	if !UseQuota {
+		return "", nil
+	}
+
+	devID, err := GetDevID(dir)
+	if err != nil {
+		return "", err
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	if mp, ok := mountPoints[devID]; ok {
+		return mp, nil
+	}
+
+	mountPoint, hasQuota, _ := quota.CheckMountpoint(devID)
+	if len(mountPoint) == 0 {
+		return mountPoint, fmt.Errorf("mountPoint not found: %s", dir)
+	}
+	if !hasQuota {
+		_, _, _, err := exec.Run(0, "mount", "-o", "remount,grpquota", mountPoint)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	vfsVersion, quotaFilename, err := quota.getVFSVersionAndQuotaFile(devID)
+	if err != nil {
+		return "", err
+	}
+
+	filename := mountPoint + "/" + quotaFilename
+	if _, err := os.Stat(filename); err != nil {
+		os.Remove(mountPoint + "/aquota.user")
+
+		header := []byte{0x27, 0x19, 0xc0, 0xd9, 0x00, 0x00, 0x00, 0x00, 0x80, 0x3a, 0x09, 0x00, 0x80,
+			0x3a, 0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x05, 0x00, 0x00, 0x00}
+		if vfsVersion == "vfsv1" {
+			header[4] = 0x01
+		}
+
+		if writeErr := ioutil.WriteFile(filename, header, 644); writeErr != nil {
+			logrus.Errorf("write file error. %s, %s, %s", filename, vfsVersion, writeErr)
+			return mountPoint, writeErr
+		}
+		if _, _, _, err := exec.Run(0, "setquota", "-g", "-t", "43200", "43200", mountPoint); err != nil {
+			os.Remove(filename)
+			return mountPoint, err
+		}
+		if err := quota.setUserQuota(0, 0, mountPoint); err != nil {
+			os.Remove(filename)
+			return mountPoint, err
+		}
+	}
+
+	// on
+	exit, stdout, stderr, err := exec.Run(0, "quotaon", "-pg", mountPoint)
+	if err != nil && exit != 1 {
```

Comment:
Should this be `if err != nil || exit != 0`?
Since if the exit code is non-zero, it means running into an error.

Reply:
exit = 1 is also valid.

Reply:
Oh, that is some kind of strange. Could you add some illustration here to tell code readers about that?